File: frontend/src/engine/math/vec2.ts
/**
 * Immutable 2D vector. Mirrors backend/app/domain/math2d/vec2.py; both
 * kernels are pinned by samples/fixtures/math-golden.json.
 */

export interface Vec2 {
  readonly x: number;
  readonly y: number;
}

export const VEC2_ZERO: Vec2 = { x: 0, y: 0 };

export function vec2(x: number, y: number): Vec2 {
  return { x, y };
}

export function add(a: Vec2, b: Vec2): Vec2 {
  return { x: a.x + b.x, y: a.y + b.y };
}

export function sub(a: Vec2, b: Vec2): Vec2 {
  return { x: a.x - b.x, y: a.y - b.y };
}

export function neg(v: Vec2): Vec2 {
  return { x: -v.x, y: -v.y };
}

export function scale(v: Vec2, factor: number): Vec2 {
  return { x: v.x * factor, y: v.y * factor };
}

export function dot(a: Vec2, b: Vec2): number {
  return a.x * b.x + a.y * b.y;
}

/** Z component of the 3D cross product; positive when `b` is CCW of `a`. */
export function cross(a: Vec2, b: Vec2): number {
  return a.x * b.y - a.y * b.x;
}

export function lengthSquared(v: Vec2): number {
  return v.x * v.x + v.y * v.y;
}

export function length(v: Vec2): number {
  return Math.hypot(v.x, v.y);
}

export function distance(a: Vec2, b: Vec2): number {
  return Math.hypot(a.x - b.x, a.y - b.y);
}

export function normalize(v: Vec2): Vec2 {
  const magnitude = length(v);
  if (magnitude <= 1e-9) {
    throw new Error("cannot normalize a zero-length vector");
  }
  return { x: v.x / magnitude, y: v.y / magnitude };
}

/** Rotate 90 degrees counterclockwise. */
export function perpendicular(v: Vec2): Vec2 {
  return { x: -v.y, y: v.x };
}

export function lerp(a: Vec2, b: Vec2, t: number): Vec2 {
  return { x: a.x + (b.x - a.x) * t, y: a.y + (b.y - a.y) * t };
}

/** Counterclockwise angle from the +X axis in degrees. */
export function angleDeg(v: Vec2): number {
  return (Math.atan2(v.y, v.x) * 180) / Math.PI;
}

export function rotateDeg(v: Vec2, degrees: number): Vec2 {
  const radians = (degrees * Math.PI) / 180;
  const cosR = Math.cos(radians);
  const sinR = Math.sin(radians);
  return { x: v.x * cosR - v.y * sinR, y: v.x * sinR + v.y * cosR };
}

export function isClose(a: Vec2, b: Vec2, tolerance = 1e-6): boolean {
  return Math.abs(a.x - b.x) <= tolerance && Math.abs(a.y - b.y) <= tolerance;
}

File: backend/app/domain/math2d/vec2.py
"""Immutable 2D vector."""

from __future__ import annotations

import math
from dataclasses import dataclass

DEFAULT_TOLERANCE = 1e-9


@dataclass(frozen=True, slots=True)
class Vec2:
    x: float
    y: float

    @staticmethod
    def zero() -> Vec2:
        return Vec2(0.0, 0.0)

    def __add__(self, other: Vec2) -> Vec2:
        return Vec2(self.x + other.x, self.y + other.y)

    def __sub__(self, other: Vec2) -> Vec2:
        return Vec2(self.x - other.x, self.y - other.y)

    def __neg__(self) -> Vec2:
        return Vec2(-self.x, -self.y)

    def scaled(self, factor: float) -> Vec2:
        return Vec2(self.x * factor, self.y * factor)

    def dot(self, other: Vec2) -> float:
        return self.x * other.x + self.y * other.y

    def cross(self, other: Vec2) -> float:
        """Z component of the 3D cross product; positive when ``other`` is CCW."""
        return self.x * other.y - self.y * other.x

    def length_squared(self) -> float:
        return self.x * self.x + self.y * self.y

    def length(self) -> float:
        return math.hypot(self.x, self.y)

    def distance_to(self, other: Vec2) -> float:
        return math.hypot(self.x - other.x, self.y - other.y)

    def normalized(self) -> Vec2:
        magnitude = self.length()
        if magnitude <= DEFAULT_TOLERANCE:
            raise ValueError("cannot normalize a zero-length vector")
        return Vec2(self.x / magnitude, self.y / magnitude)

    def perpendicular(self) -> Vec2:
        """Rotate 90 degrees counterclockwise."""
        return Vec2(-self.y, self.x)

    def lerp(self, other: Vec2, t: float) -> Vec2:
        return Vec2(self.x + (other.x - self.x) * t, self.y + (other.y - self.y) * t)

    def angle_deg(self) -> float:
        """Counterclockwise angle from the +X axis in degrees."""
        return math.degrees(math.atan2(self.y, self.x))

    def rotated_deg(self, degrees: float) -> Vec2:
        radians = math.radians(degrees)
        cos_r = math.cos(radians)
        sin_r = math.sin(radians)
        return Vec2(self.x * cos_r - self.y * sin_r, self.x * sin_r + self.y * cos_r)

    def is_close(self, other: Vec2, tolerance: float = 1e-6) -> bool:
        return abs(self.x - other.x) <= tolerance and abs(self.y - other.y) <= tolerance
